openedx/core/djangoapps/programs/utils.py

```diff
@@ -461,57 +462,70 @@ def _attach_course_run_upgrade_url(self, run_mode):
+
+    def _filter_out_courses_with_enrollments(self, courses):
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
         )
+        course_runs_with_enrollments = set(unicode(enrollment.course_id) for enrollment in enrollments)
+        courses_without_enrollments = []
+        for course in courses:
+            if len([run for run in course['course_runs'] if unicode(run['key']) in course_runs_with_enrollments]) == 0:
+                courses_without_enrollments.append(course)
 
-        is_paid_seat = False
-        if enrollment_mode is not None and active is not None and active is True:
-            # Check all the applicable seat types
-            # this will also check for no-id-professional as professional
-            is_paid_seat = any(seat_type in enrollment_mode for seat_type in applicable_seat_types)
-
-        return is_paid_seat
+        return courses_without_enrollments
 
     def _collect_one_click_purchase_eligibility_data(self):
         """
         Extend the program data with data about learner's eligibility for one click purchase,
         discount data of the program and SKUs of seats that should be added to basket.
         """
-        applicable_seat_types = self.data['applicable_seat_types']
+        applicable_seat_types = set(seat for seat in self.data['applicable_seat_types'] if seat != 'credit')
         is_learner_eligible_for_one_click_purchase = self.data['is_program_eligible_for_one_click_purchase']
         skus = []
         bundle_variant = 'full'
-        if is_learner_eligible_for_one_click_purchase:
-            for course in self.data['courses']:
-                add_course_sku = True
-                course_runs = course.get('course_runs', [])
-                published_course_runs = filter(lambda run: run['status'] == 'published', course_runs)
-
-                if len(published_course_runs) == 1:
-                    for course_run in course_runs:
-                        is_paid_seat = self._check_enrollment_for_user(course_run)
 
-                        if is_paid_seat:
-                            add_course_sku = False
-                            break
-
-                    if add_course_sku:
+        if is_learner_eligible_for_one_click_purchase:
+            # Reduce the set of required courses to those for which the user doesn't have an active enrollment
+            courses = self._filter_out_courses_with_enrollments(self.data['courses'])
+
+            # Reduce the set of required courses to those for which the user doesn't have an entitlement
+            courses = self._filter_out_courses_with_entitlements(courses)
+
+            if len(courses) < len(self.data['courses']):
+                bundle_variant = 'partial'
+
+            for course in courses:
+                entitlements = []
+                for entitlement in course['entitlements']:
```

Comment:
Nit: I recommend using .get() rather than accessing the dictionary in this fashion.  With .get() you can return a default value and avoid exceptions for missing data. Consider it precautionary coding.

Reply:
If a course has no entitlements the field will still appear in the api response, with an empty list as its associated value.